### PR TITLE
CO and PNW

### PR DESCRIPTION
**Pull request by dblodgett-usgs**
image
Current workflow graph.

### CODE REVIEW

File: _targets.R [hunk 1]
```diff
@@ -62,40 +75,60 @@ list(
...
-  # that contains that primary key. 
-  tar_target("providers", read_csv("reg/providers.csv")),
-  
-  
-  tar_target("registry", build_registry(gage_locations,
-                                        registry = "reg/ref_gages.csv",
-                                        providers = providers)),
+  # based on all known gages from all providers, find potential duplicates.
+  # starts by finding gages within 100m of eachother then checks if they are on different flowlines.
+  tar_target("duplicate_locations", find_duplicate_locations(gage_hydrologic_locations_with_mainstems)),
```
Comment (dblodgett-usgs) [inline _targets.R:122]: This now bruit force finds gages that are very close together.

File: _targets.R [hunk 2]
```diff
@@ -62,40 +75,60 @@ list(
...
-                                                        registry, providers, usgs_reference_file, 
+  tar_target("usgs_reference_out", write_usgs_reference(gage_hydrologic_locations_with_mainstems,
+                                                        registry, providers, usgs_reference_file,
                                                         usgs_nldi_file)),
   
   tar_target("reference_out", write_reference(gage_hydrologic_locations_with_mainstems, 
                                               registry, providers, reference_file, 
-                                              nldi_file)),
+                                              nldi_file,
+                                              duplicate_locations = duplicate_locations)),
```
Comment (dblodgett-usgs) [inline _targets.R:133]: duplicate locations get added as a comma seperated list so you can see that gages are very close to eachother.

File: _targets.R [hunk 3]
```diff
@@ -62,40 +75,60 @@ list(
...
+  tar_target("registry", build_registry(gage_locations,
+                                        registry = "reg/ref_gages.csv",
+                                        providers = providers)),
+  
+  # Also create a table of reference locations for the registered gages.
+  # unlike the registry, this may update to have the "best" location of a gage. 
+  tar_target("ref_locations", build_reference_location(gage_locations, 
+                                                       reference_locations = "reg/ref_locations.csv", 
+                                                       registry = "reg/ref_gages.csv", 
+                                                       providers = providers)),
```
Comment (dblodgett-usgs) [inline _targets.R:97]: This "ref_locations" is now checked in as a csv of the lat/lon of every location that's ever been seen from every provider.

File: _targets.R [hunk 4]
```diff
@@ -62,40 +75,60 @@ list(
...
+  # Each entry will have a provider and provider_id that acts as a unique
+  # primary key. The existing registry file will have a unique attribute
+  # that contains that primary key. 
+  tar_target("providers_csv", command = "reg/providers.csv", format = "file"),
+  tar_target("providers", read_csv(providers_csv)),
+  
+  
+  tar_target("registry", build_registry(gage_locations,
+                                        registry = "reg/ref_gages.csv",
+                                        providers = providers)),
```
Comment (dblodgett-usgs) [inline _targets.R:90]: Much of this was just reorganized so the workflow file is more linear.

**Comment by ksonda** [comment]
Looks good to me

File: R/gage_locations.R [hunk 5]
```diff
@@ -9,45 +11,114 @@ get_gage_locations <- function(nwis_gage, streamstats_sites, cdec_gage) {
...
+              by = c("org" = "Organization")) |>
+    mutate(description = paste0(org, " Streamflow Site")) |>
+    mutate(description = ifelse(!`stream type` %in% c("unknown", "NA"),
+                                paste0(description, " Type: ", `stream type`),
+                                description)) |>
+    select(name = Site_Name,
+           description, 
+           subjectOf = url,
+           provider = provider_url,
+           provider_id = org_SiteNo)
```
Comment (dblodgett-usgs) [inline R/gage_locations.R:97]: I just went through and cleaned up the pnw_gage organizations and picked out the organizations that have what appears to be the best quality data in the set. 

Would either @ksonda or @ewiggansLI be willing to look over the attached? 

`pnw_gages.csv` is filtered per the code here. 
`all_pnw_gages.csv` is not filtered and I've included all the variables available. 
`pnw_provider.csv` is the `p_gage_provider` variable from the code above. 

Question is, are there providers I should add or remove at this stage? Do you see anything amiss in the data?

[pnw_gages.csv](https://github.com/internetofwater/ref_gages/files/11730173/pnw_gages.csv)
[all_pnw_gage.csv](https://github.com/internetofwater/ref_gages/files/11730186/all_pnw_gage.csv)
[pnw_provider.csv](https://github.com/internetofwater/ref_gages/files/11730188/pnw_provider.csv)

Comment (ewiggansLI) [reply R/gage_locations.R:97]: Will be using the pnw_gages.csv to create items in Geoconnex. Thanks!

Comment (webb-ben) [reply R/gage_locations.R:97]: @dblodgett-usgs should provider_id in `pnw_gages.csv` be unique? There are an awful lot that are `NA` or `n/a` which makes the creation of a uri less straightforward.

Comment (dblodgett-usgs) [reply R/gage_locations.R:97]: I kind of left off not knowing what to do on that point which is why this has sat unfinished for so long...

**Pull request merged**